Hi,

I am evidently missing something. The [Travis build](https://travis-ci.org/realpython/flask-registration/builds/155100440) gives the following:

```
$ python manage.py test
Traceback (most recent call last):
  File "manage.py", line 19, in <module>
    from project import app, db
  File "/home/travis/build/realpython/flask-registration/project/__init__.py", line 24, in <module>
    app.config.from_object(os.environ['APP_SETTINGS'])
  File "/home/travis/virtualenv/python2.7.12/lib/python2.7/UserDict.py", line 40, in __getitem__
    raise KeyError(key)
KeyError: 'APP_SETTINGS'
```

That error is raised by [L24](https://github.com/realpython/flask-registration/blob/master/project/__init__.py#L24) in `__init__.py` which is a file not touched by this PR. That operation is executed before anything I could have modified is called or executed, at least as far as I know.

I have (locally) the same error if I do not set `APP_SETTINGS`, and I have the same error if I use the code from your master branch (i. e. https://github.com/realpython/flask-registration/commit/db371d6134c4ac93c8b86795c4575c9fd3dd6f7a)

Furthermore, the `KeyError` means that the env variable `APP_SETTINGS` is not present in the dictionary, i.e. that the env variable is not set.

What am I missing?

--- a/CONTRIBUTING.md
+++ b/CONTRIBUTING.md
@@ -1,0 +1,101 @@
+How to contribute
+-----------------
+
+## env.sh
+
+You can create a file `env.sh` with the settings needed to setup the needed development (or testing) environment variables.
+
+```bash
+#!/usr/bin/env bash
+
+function app_clean_env() {
+    unset APP_MAIL_SERVER
+    unset APP_MAIL_PORT
+    unset APP_MAIL_USE_TLS
+    unset APP_MAIL_USE_SSL
+    unset APP_MAIL_USERNAME
+    unset APP_MAIL_PASSWORD
+    unset APP_MAIL_DEFAULT_SENDER
+    unset APP_SQLALCHEMY_DATABASE_URI
+}
+
+app_clean_env
+
+if [[ "$APP_SETTINGS" == 'project.config.DevelopmentConfig' ]]; then
+    echo "Apply DevelopmentConfig settings"
+
+    # mail setting
+    export APP_MAIL_SERVER='<your_smtp_server>'
+    export APP_MAIL_PORT='<your_smtp_server_port>'
+    export APP_MAIL_USE_TLS='<true_or_false>'
+    export APP_MAIL_USE_SSL='<true_or_false>'
+
+    # mail authentication
+    export APP_MAIL_USERNAME='<your_smtp_username>'
+    export APP_MAIL_PASSWORD='<your_smtp_password>'
+
+elif [[ "$APP_SETTINGS" == 'project.config.TestingConfig' ]]; then
+    echo "Apply TestingConfig settings"
+    # put your testing settings here
+
+elif [[ "$APP_SETTINGS" == 'project.config.ProductionConfig' ]]; then
+    echo "Apply ProductionConfig settings"
+    # put your production settings here
+
+else
+    (>&2 echo "Unrecognized setting")
+fi
+```
+
+## Sending and debugging emails
+
+To send email for development, testing and production you can use your own hosted SMTP server or other solutions. Here's two examples:
+
+* [Mailgun](https://mailgun.com)
+You can use Mailgun if you want to send emails in production.
+
+* [Debug mail](https://mailgun.com)
+Debug mail is a tool to debut email, you can use it for developing testing.
+
+Both services offer free plans that you can use for developing and testing `flask-registration`.
+
+## Develop locally
+
+You can use the scripts `clean.sh` and `create.sh` to setup the environment.
+
+You need to set up the `APP_SETTINGS` variable:
+```bash
+export APP_SETTINGS=project.config.DevelopmentConfig
+```
+
+To run the development server:
+```bash
+./clean.sh; source env.sh && ./create.sh && python manage.py runserver
+```
+
+To run the tests:
+```bash
+./clean.sh; source env.sh && python manage.py test
+```
+
+## Use Travis to run tests
+
+1. Fork the repository
+2. if you don't have an account on [Travis](http://travis-ci.org/), create it (it's free)
+3. Add your `flask-registration` repository on the Travis settings page.
+4. In Travis, click of the cog icon next to the repository and add the following [environment variables](https://docs.travis-ci.com/user/environment-variables/#Defining-Variables-in-Repository-Settings):
+```
+APP_SETTINGS: project.config.TestingConfig
+APP_MAIL_SERVER: <your_smtp_server>
+APP_MAIL_PORT: <your_smtp_server_port>
+APP_MAIL_USERNAME: <your_smpt_username>
+APP_MAIL_PASSWORD: <your_smpt_password>
+```
+you can use any valid SMTP credentials, the defaults are the following:
+```
+MAIL_SERVER = 'smtp.googlemail.com'
+MAIL_PORT = 465
+MAIL_USE_TLS = False
+MAIL_USE_SSL = True
+```
+when the commits to your repository, travis will automatically build the application with the testing settings (see the file `.travis.yml`.

--- a/clean.sh
+++ b/clean.sh
@@ -1,0 +1,18 @@
+#!/usr/bin/env bash
+
+function clean_env() {
+    unset APP_MAIL_SERVER
+    unset APP_MAIL_PORT
+    unset APP_MAIL_USE_TLS
+    unset APP_MAIL_USE_SSL
+    unset APP_MAIL_USERNAME
+    unset APP_MAIL_PASSWORD
+    unset APP_MAIL_DEFAULT_SENDER
+    unset APP_SQLALCHEMY_DATABASE_URI
+}
+
+clean_env
+
+rm -rf migrations
+rm -rf tmp
+rm -f project/dev.sqlite

--- a/create.sh
+++ b/create.sh
@@ -1,0 +1,5 @@
+#!/usr/bin/env bash
+
+python manage.py create_db
+python manage.py db init
+python manage.py db migrate

--- a/project/config.py
+++ b/project/config.py
@@ -4,6 +4,20 @@ import os
 basedir = os.path.abspath(os.path.dirname(__file__))
 
 
+def _get_bool_env_var(varname, default=None):
+
+    value = os.environ.get(varname, default)
+
+    if value is None:
+        return False
+    elif isinstance(value, str) and value.lower() == 'false':
+        return False
+    elif bool(value) is False:
+        return False
+    else:
+        return bool(value)
+
+
 class BaseConfig(object):
     """Base configuration."""
 
@@ -17,12 +31,17 @@ class BaseConfig(object):
     DEBUG_TB_INTERCEPT_REDIRECTS = False
 
     # mail settings
-    MAIL_SERVER = 'smtp.googlemail.com'
-    MAIL_PORT = 465
-    MAIL_USE_TLS = False
-    MAIL_USE_SSL = True
+    # defaults are:
+    #  - MAIL_SERVER = 'smtp.googlemail.com'
+    #  - MAIL_PORT = 465
+    #  - MAIL_USE_TLS = False
+    #  - MAIL_USE_SSL = True
+    MAIL_SERVER = os.environ.get('APP_MAIL_SERVER', 'smtp.googlemail.com')
+    MAIL_PORT = int(os.environ.get('APP_MAIL_PORT', 465))
+    MAIL_USE_TLS = _get_bool_env_var('APP_MAIL_USE_TLS', False)
+    MAIL_USE_SSL = _get_bool_env_var('APP_MAIL_USE_SSL', True)
 
-    # gmail authentication
+    # mail authentication
     MAIL_USERNAME = os.environ['APP_MAIL_USERNAME']
     MAIL_PASSWORD = os.environ['APP_MAIL_PASSWORD']
 
@@ -40,6 +59,7 @@ class DevelopmentConfig(BaseConfig):
 
 class TestingConfig(BaseConfig):
     """Testing configuration."""
+    LOGIN_DISABLED=False
     TESTING = True
     DEBUG = False
     BCRYPT_LOG_ROUNDS = 1

--- a/project/main/views.py
+++ b/project/main/views.py
@@ -5,9 +5,10 @@
 #### imports ####
 #################
 
-from flask import render_template, Blueprint
-from flask.ext.login import login_required
-
+from flask import render_template
+from flask import Blueprint
+from flask_login import current_user
+from flask_login import login_required
 
 ################
 #### config ####
@@ -21,6 +22,5 @@ main_blueprint = Blueprint('main', __name__,)
 ################
 
 @main_blueprint.route('/')
-@login_required
 def home():
-    return render_template('main/index.html')
+    return render_template('main/index.html', current_user=current_user)

--- a/project/models.py
+++ b/project/models.py
@@ -17,15 +17,18 @@ class User(db.Model):
     admin = db.Column(db.Boolean, nullable=False, default=False)
     confirmed = db.Column(db.Boolean, nullable=False, default=False)
     confirmed_on = db.Column(db.DateTime, nullable=True)
+    password_reset_token = db.Column(db.String, nullable=True)
 
     def __init__(self, email, password, confirmed,
-                 admin=False, confirmed_on=None):
+                 admin=False, confirmed_on=None,
+                 password_reset_token=None):
         self.email = email
         self.password = bcrypt.generate_password_hash(password)
         self.registered_on = datetime.datetime.now()
         self.admin = admin
         self.confirmed = confirmed
         self.confirmed_on = confirmed_on
+        self.password_reset_token = password_reset_token
 
     def is_authenticated(self):
         return True

--- a/project/templates/main/index.html
+++ b/project/templates/main/index.html
@@ -1,8 +1,25 @@
 {% extends "_base.html" %}
 {% block content %}
 
-<h1>Welcome!</h1>
-<br>
-<p><a href="/logout">Logout</a>
+{% if current_user.is_authenticated() %}
+  <h1>Welcome, {{ current_user.email }}</h1>
+{% else %}
+  <h1>Welcome!</h1>
+{% endif %}
+<br />
+
+{% if current_user.is_authenticated() %}
+  <p>You are <span style="color:green">authenticated</span></p>
+  <ul>
+    <li><a href="/logout">Logout</a></li>
+  </ul>
+{% else %}
+  <p>You are <span style="color:red">not authenticated</span>:</p>
+  <ul>
+    <li><a href="/login">Sign In</a></li>
+    <li>Need to <a href="/register">register</a>?</li>
+    <li><a href="/forgot">Forgot</a> your password?</li>
+  </ul>
+{% endif %}
 
 {% endblock %}

--- a/project/templates/user/forgot.html
+++ b/project/templates/user/forgot.html
@@ -1,0 +1,24 @@
+{% extends "_base.html" %}
+
+{% block content %}
+
+<h1>Forgot Your Password?</h1>
+<br />
+<form class="form-signin" role="form" method="post" action="">
+  {{ form.csrf_token }}
+  {{ form.email(placeholder="email") }}
+    <span class="error">
+      {% if form.email.errors %}
+        {% for error in form.email.errors %}
+          {{ error }}
+        {% endfor %}
+      {% endif %}
+    </span>
+  </p>
+  <button class="btn btn-success" type="submit">Reset your password</button>
+</form>
+<ul>
+<li>Back to <a href="/login">login</a></li>
+<li>Need to <a href="/register">register</a>?</li>
+</ul>
+{% endblock %}

--- a/project/templates/user/forgot_new.html
+++ b/project/templates/user/forgot_new.html
@@ -1,0 +1,33 @@
+{% extends "_base.html" %}
+
+{% block content %}
+
+<h1>Forgot Your Password?</h1>
+<br />
+<form class="form-signin" role="form" method="post" action="">
+  {{ form.csrf_token }}
+  <p>
+  {{ form.password(placeholder="password") }}
+    <span class="error">
+      {% if form.password.errors %}
+        {% for error in form.password.errors %}
+          {{ error }}
+        {% endfor %}
+      {% endif %}
+    </span>
+  </p>
+  <p>
+  {{ form.confirm(placeholder="confirm") }}
+    <span class="error">
+      {% if form.confirm.errors %}
+        {% for error in form.confirm.errors %}
+          {{ error }}
+        {% endfor %}
+      {% endif %}
+    </span>
+  </p>
+  <button class="btn btn-success" type="submit">Reset your password!</button>
+  <br /><br />
+</form>
+
+{% endblock %}

--- a/project/templates/user/login.html
+++ b/project/templates/user/login.html
@@ -27,8 +27,11 @@
     </span>
   </p>
   <button class="btn btn-success" type="submit">Sign in!</button>
-  <br><br>
-  <p>Need to <a href="/register">Register</a>?</p>
+  <br /><br />
+  <ul>
+  <li>Need to <a href="/register">Register</a>?</li>
+  <li><a href="/forgot">Forgot</a> your password?</li>
+  </ul>
 </form>
 
 {% endblock %}

--- a/project/templates/user/reset.html
+++ b/project/templates/user/reset.html
@@ -1,0 +1,6 @@
+<p>Hi! Somobody requested a password reset for the account {{ username }}.</p>
+<p>Click on the following link to reset your password:</p>
+<p><a href="{{ reset_url }}">{{ reset_url }}</a></p>
+<p>If you didn't request it, you can ignore this email.</p>
+<br />
+<p>Cheers!</p>

--- a/project/user/forms.py
+++ b/project/user/forms.py
@@ -16,10 +16,10 @@ class LoginForm(Form):
 class RegisterForm(Form):
     email = TextField(
         'email',
-        validators=[DataRequired(), Email(message=None), Length(min=6, max=40)])
+        validators=[DataRequired(), Email(message=None), Length(min=6, max=255)])
     password = PasswordField(
         'password',
-        validators=[DataRequired(), Length(min=6, max=25)]
+        validators=[DataRequired(), Length(min=6, max=255)]
     )
     confirm = PasswordField(
         'Repeat password',
@@ -40,10 +40,26 @@ class RegisterForm(Form):
         return True
 
 
+class ForgotForm(Form):
+    email = TextField(
+        'email',
+        validators=[DataRequired(), Email(message=None), Length(min=6, max=255)])
+
+    def validate(self):
+        initial_validation = super(ForgotForm, self).validate()
+        if not initial_validation:
+            return False
+        user = User.query.filter_by(email=self.email.data).first()
+        if not user:
+            self.email.errors.append("This email is not registered")
+            return False
+        return True
+
+
 class ChangePasswordForm(Form):
     password = PasswordField(
         'password',
-        validators=[DataRequired(), Length(min=6, max=25)]
+        validators=[DataRequired(), Length(min=6, max=255)]
     )
     confirm = PasswordField(
         'Repeat password',

--- a/project/user/views.py
+++ b/project/user/views.py
@@ -9,7 +9,7 @@ import datetime
 
 from flask import render_template, Blueprint, url_for, \
     redirect, flash, request
-from flask.ext.login import login_user, logout_user, \
+from flask_login import login_user, logout_user, \
     login_required, current_user
 
 from project.models import User
@@ -17,7 +17,7 @@ from project.email import send_email
 from project.token import generate_confirmation_token, confirm_token
 from project.decorators import check_confirmed
 from project import db, bcrypt
-from .forms import LoginForm, RegisterForm, ChangePasswordForm
+from .forms import LoginForm, RegisterForm, ChangePasswordForm, ForgotForm
 
 
 ################
@@ -137,3 +137,58 @@ def resend_confirmation():
     send_email(current_user.email, subject, html)
     flash('A new confirmation email has been sent.', 'success')
     return redirect(url_for('user.unconfirmed'))
+
+@user_blueprint.route('/forgot',  methods=['GET', 'POST'])
+def forgot():
+    form = ForgotForm(request.form)
+    if form.validate_on_submit():
+
+        user = User.query.filter_by(email=form.email.data).first()
+        token = generate_confirmation_token(user.email)
+
+        user.password_reset_token = token
+        db.session.commit()
+
+        reset_url = url_for('user.forgot_new', token=token, _external=True)
+        html = render_template('user/reset.html',
+                               username=user.email,
+                               reset_url=reset_url)
+        subject = "Reset your password"
+        send_email(user.email, subject, html)
+
+        flash('A password reset email has been sent via email.', 'success')
+        return redirect(url_for("main.home"))
+
+    return render_template('user/forgot.html', form=form)
+
+
+@user_blueprint.route('/forgot/new/<token>', methods=['GET', 'POST'])
+def forgot_new(token):
+
+    email = confirm_token(token)
+    user = User.query.filter_by(email=email).first_or_404()
+
+    if user.password_reset_token is not None:
+        form = ChangePasswordForm(request.form)
+        if form.validate_on_submit():
+            user = User.query.filter_by(email=email).first()
+            if user:
+                user.password = bcrypt.generate_password_hash(form.password.data)
+                user.password_reset_token = None
+                db.session.commit()
+
+                login_user(user)
+
+                flash('Password successfully changed.', 'success')
+                return redirect(url_for('user.profile'))
+
+            else:
+                flash('Password change was unsuccessful.', 'danger')
+                return redirect(url_for('user.profile'))
+        else:
+            flash('You can now change your password.', 'success')
+            return render_template('user/forgot_new.html', form=form)
+    else:
+        flash('Can not reset the password, try again.', 'danger')
+
+    return redirect(url_for('main.home'))

--- a/tests/test_main.py
+++ b/tests/test_main.py
@@ -5,11 +5,11 @@ from project.util import BaseTestCase
 
 class TestMainViews(BaseTestCase):
 
-    def test_main_route_requires_login(self):
+    def test_main_route_does_not_require_login(self):
         # Ensure main route requres a logged in user.
         response = self.client.get('/', follow_redirects=True)
         self.assertTrue(response.status_code == 200)
-        self.assertTemplateUsed('user/login.html')
+        self.assertTemplateUsed('main/index.html')
 
 
 if __name__ == '__main__':

--- a/tests/test_user.py
+++ b/tests/test_user.py
@@ -7,7 +7,7 @@ from project import db
 from project.models import User
 from project.util import BaseTestCase
 from project.user.forms import RegisterForm, \
-    LoginForm, ChangePasswordForm
+    LoginForm, ChangePasswordForm, ForgotForm
 from project.token import generate_confirmation_token, confirm_token
 
 
@@ -59,6 +59,21 @@ class TestUserForms(BaseTestCase):
     def test_validate_invalid_change_password_format(self):
         # Ensure invalid email format throws error.
         form = ChangePasswordForm(password='123', confirm='123')
+        self.assertFalse(form.validate())
+
+    def test_validate_success_forgot_password(self):
+        # Ensure invalid email format throws error.
+        form = ForgotForm(email='test@user.com')
+        self.assertTrue(form.validate())
+
+    def test_validate_invalid_forgot_password_format(self):
+        # Ensure invalid email format throws error.
+        form = ForgotForm(email='unknown')
+        self.assertFalse(form.validate())
+
+    def test_validate_invalid_forgot_password_no_such_user(self):
+        # Ensure invalid email format throws error.
+        form = ForgotForm(email='not@correct.com')
         self.assertFalse(form.validate())
 
 
@@ -139,6 +154,115 @@ class TestUserViews(BaseTestCase):
         db.session.commit()
         token = generate_confirmation_token('test@test1.com')
         self.assertFalse(confirm_token(token, -1))
+
+    def test_forgot_password_does_not_require_login(self):
+        # Ensure user can request new password without login.
+        self.client.get('/forgot', follow_redirects=True)
+        self.assertTemplateUsed('user/forgot.html')
+
+    def test_correct_forgot_password_request(self):
+        # Ensure login behaves correctly with correct credentials.
+        with self.client:
+            response = self.client.post(
+                '/forgot',
+                data=dict(email="test@user.com"),
+                follow_redirects=True
+            )
+            self.assertTrue(response.status_code == 200)
+            self.assertTemplateUsed('main/index.html')
+
+    def test_reset_forgotten_password_valid_token(self):
+        # Ensure user can confirm account with valid token.
+        with self.client:
+            self.client.post('/forgot', data=dict(
+                email='test@user.com',
+            ), follow_redirects=True)
+            token = generate_confirmation_token('test@user.com')
+            response = self.client.get('/forgot/new/'+token, follow_redirects=True)
+            self.assertTemplateUsed('user/forgot_new.html')
+            self.assertIn(
+                b'You can now change your password.',
+                response.data
+            )
+            self.assertFalse(current_user.is_authenticated())
+
+    def test_reset_forgotten_password_valid_token_correct_login(self):
+        # Ensure user can confirm account with valid token.
+        with self.client:
+            self.client.post('/forgot', data=dict(
+                email='test@user.com',
+            ), follow_redirects=True)
+            token = generate_confirmation_token('test@user.com')
+            response = self.client.get('/forgot/new/'+token, follow_redirects=True)
+            self.assertTemplateUsed('user/forgot_new.html')
+            self.assertIn(
+                b'You can now change your password.',
+                response.data
+            )
+            response = self.client.post(
+                '/forgot/new/'+token,
+                data=dict(password="new-password", confirm="new-password"),
+                follow_redirects=True
+            )
+            self.assertIn(
+                b'Password successfully changed.',
+                response.data
+            )
+            self.assertTemplateUsed('user/profile.html')
+            self.assertTrue(current_user.is_authenticated())
+            self.client.get('/logout')
+            self.assertFalse(current_user.is_authenticated())
+
+            response = self.client.post(
+                '/login',
+                data=dict(email="test@user.com", password="new-password"),
+                follow_redirects=True
+            )
+            self.assertTrue(response.status_code == 200)
+            self.assertTrue(current_user.email == "test@user.com")
+            self.assertTrue(current_user.is_active())
+            self.assertTrue(current_user.is_authenticated())
+            self.assertTemplateUsed('main/index.html')
+
+    def test_reset_forgotten_password_valid_token_invalid_login(self):
+        # Ensure user can confirm account with valid token.
+        with self.client:
+            self.client.post('/forgot', data=dict(
+                email='test@user.com',
+            ), follow_redirects=True)
+            token = generate_confirmation_token('test@user.com')
+            response = self.client.get('/forgot/new/'+token, follow_redirects=True)
+            self.assertTemplateUsed('user/forgot_new.html')
+            self.assertIn(
+                b'You can now change your password.',
+                response.data
+            )
+            response = self.client.post(
+                '/forgot/new/'+token,
+                data=dict(password="new-password", confirm="new-password"),
+                follow_redirects=True
+            )
+            self.assertIn(
+                b'Password successfully changed.',
+                response.data
+            )
+            self.assertTemplateUsed('user/profile.html')
+            self.assertTrue(current_user.is_authenticated())
+            self.client.get('/logout')
+            self.assertFalse(current_user.is_authenticated())
+
+            response = self.client.post(
+                '/login',
+                data=dict(email="test@user.com", password="just_a_test_user"),
+                follow_redirects=True
+            )
+            self.assertTrue(response.status_code == 200)
+            self.assertFalse(current_user.is_authenticated())
+            self.assertIn(
+                b'Invalid email and/or password.',
+                response.data
+            )
+            self.assertTemplateUsed('user/login.html')
 
 
 if __name__ == '__main__':

